[netatmo] Thing is displayed as online but the device is offline

I have a Presence camera which is currently not power connected, but the thing is displayed as online.

I would expect that it is displayed as offline, when it isn't available.

Possible solution: The Netatmo /gethomedata endpoint returns the camera with "status" : "disconnected" instead of "on". I would expect that this information is used to determine the thing status.

I just discovered we have an exception when the camera is off:
```
16:56:58.219 [ERROR] [internal.items.ItemStateConverterImpl] - A conversion of null was requested:
java.lang.IllegalArgumentException: State must not be null.
        at org.eclipse.smarthome.core.internal.items.ItemStateConverterImpl.convertToAcceptedState(ItemStateConverterImpl.java:58) ~[?:?]
        at org.eclipse.smarthome.core.thing.internal.profiles.ProfileCallbackImpl.sendUpdate(ProfileCallbackImpl.java:134) ~[?:?]
        at org.eclipse.smarthome.core.thing.internal.profiles.SystemDefaultProfile.onStateUpdateFromHandler(SystemDefaultProfile.java:53) ~[?:?]
        at org.eclipse.smarthome.core.thing.internal.CommunicationManager.lambda$9(CommunicationManager.java:467) ~[?:?]
        at org.eclipse.smarthome.core.thing.internal.CommunicationManager.lambda$11(CommunicationManager.java:487) ~[?:?]
        at java.lang.Iterable.forEach(Iterable.java:75) ~[?:1.8.0_152]
        at org.eclipse.smarthome.core.thing.internal.CommunicationManager.handleCallFromHandler(CommunicationManager.java:483) ~[?:?]
        at org.eclipse.smarthome.core.thing.internal.CommunicationManager.stateUpdated(CommunicationManager.java:465) ~[?:?]
        at org.eclipse.smarthome.core.thing.internal.ThingManagerImpl$1.stateUpdated(ThingManagerImpl.java:168) ~[?:?]
        at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateState(BaseThingHandler.java:245) ~[?:?]
        at org.openhab.binding.netatmo.internal.handler.AbstractNetatmoThingHandler.lambda$3(AbstractNetatmoThingHandler.java:156) ~[?:?]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184) ~[?:1.8.0_152]
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175) ~[?:1.8.0_152]
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382) ~[?:1.8.0_152]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[?:1.8.0_152]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[?:1.8.0_152]
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151) ~[?:1.8.0_152]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174) ~[?:1.8.0_152]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:1.8.0_152]
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418) ~[?:1.8.0_152]
        at org.openhab.binding.netatmo.internal.handler.AbstractNetatmoThingHandler.updateDataChannels(AbstractNetatmoThingHandler.java:151) ~[?:?]
        at org.openhab.binding.netatmo.internal.handler.AbstractNetatmoThingHandler.updateChannels(AbstractNetatmoThingHandler.java:144) ~[?:?]
        at org.openhab.binding.netatmo.internal.handler.NetatmoModuleHandler.updateChannels(NetatmoModuleHandler.java:106) ~[?:?]
        at org.openhab.binding.netatmo.internal.handler.NetatmoDeviceHandler.lambda$9(NetatmoDeviceHandler.java:209) ~[?:?]
        at java.util.Optional.ifPresent(Optional.java:159) ~[?:1.8.0_152]
        at org.openhab.binding.netatmo.internal.handler.NetatmoDeviceHandler.lambda$7(NetatmoDeviceHandler.java:207) ~[?:?]
        at java.util.concurrent.ConcurrentHashMap.forEach(ConcurrentHashMap.java:1597) ~[?:1.8.0_152]
        at org.openhab.binding.netatmo.internal.handler.NetatmoDeviceHandler.updateChildModules(NetatmoDeviceHandler.java:206) ~[?:?]
        at org.openhab.binding.netatmo.internal.handler.NetatmoDeviceHandler.updateChannels(NetatmoDeviceHandler.java:164) ~[?:?]
        at org.openhab.binding.netatmo.internal.handler.NetatmoDeviceHandler.lambda$0(NetatmoDeviceHandler.java:83) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_152]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_152]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_152]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_152]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_152]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_152]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_152]
```
When adding null annotations, I suppressed a test againt state being null, this was probably a bad idea but I don't understand how the state can be null. To be investigated and identify the channel for which we determine a null state.

I can confirm that the state of the camera switched to "disconnected" after few minutes.

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/ChannelTypeUtils.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/ChannelTypeUtils.java
@@ -25,9 +25,11 @@ import org.eclipse.smarthome.core.library.types.DateTimeType;
 import org.eclipse.smarthome.core.library.types.DecimalType;
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.library.types.QuantityType;
+import org.eclipse.smarthome.core.library.types.RawType;
 import org.eclipse.smarthome.core.library.types.StringType;
 import org.eclipse.smarthome.core.types.State;
 import org.eclipse.smarthome.core.types.UnDefType;
+import org.eclipse.smarthome.io.net.http.HttpUtil;
 
 /**
  * This class holds various channel values conversion methods
@@ -126,5 +128,10 @@ public class ChannelTypeUtils {
 
     public static State toQuantityType(@Nullable BigDecimal value, Unit<?> unit) {
         return value == null ? UnDefType.NULL : new QuantityType<>(value, unit);
+    }
+
+    public static State toRawType(String pictureUrl) {
+        RawType picture = HttpUtil.downloadImage(pictureUrl);
+        return picture == null ? UnDefType.UNDEF : picture;
     }
 }

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/camera/CameraHandler.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/camera/CameraHandler.java
@@ -12,7 +12,7 @@
  */
 package org.openhab.binding.netatmo.internal.camera;
 
-import static org.openhab.binding.netatmo.internal.ChannelTypeUtils.toOnOffType;
+import static org.openhab.binding.netatmo.internal.ChannelTypeUtils.*;
 import static org.openhab.binding.netatmo.internal.NetatmoBindingConstants.*;
 
 import java.io.IOException;
@@ -123,7 +123,7 @@ public abstract class CameraHandler extends NetatmoModuleHandler<NAWelcomeCamera
 
     protected State getLivePictureState() {
         Optional<String> livePictureURL = getLivePictureURL();
-        return livePictureURL.isPresent() ? HttpUtil.downloadImage(livePictureURL.get()) : UnDefType.UNDEF;
+        return livePictureURL.isPresent() ? toRawType(livePictureURL.get()) : UnDefType.UNDEF;
     }
 
     protected State getLiveStreamState() {
@@ -236,5 +236,11 @@ public abstract class CameraHandler extends NetatmoModuleHandler<NAWelcomeCamera
             logger.warn("Error on accessing local camera url!", e);
         }
         return Optional.empty();
+    }
+
+    @Override
+    protected boolean isReachable() {
+        Optional<NAWelcomeCamera> module = getModule();
+        return module.isPresent() ? !"disconnected".equalsIgnoreCase(module.get().getStatus()) : false;
     }
 }

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/AbstractNetatmoThingHandler.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/AbstractNetatmoThingHandler.java
@@ -141,6 +141,10 @@ public abstract class AbstractNetatmoThingHandler extends BaseThingHandler {
     }
 
     protected void updateChannels() {
+        if (thing.getStatus() != ThingStatus.ONLINE) {
+            return;
+        }
+
         updateDataChannels();
 
         triggerEventChannels();
@@ -277,5 +281,9 @@ public abstract class AbstractNetatmoThingHandler extends BaseThingHandler {
             channels.add(channel);
             types.add(type);
         }
+    }
+
+    protected boolean isReachable() {
+        return true;
     }
 }

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/NetatmoDeviceHandler.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/NetatmoDeviceHandler.java
@@ -136,10 +136,10 @@ public abstract class NetatmoDeviceHandler<DEVICE> extends AbstractNetatmoThingH
                                 "Unable to connect Netatmo API : " + e.getLocalizedMessage());
                     }
                     if (newDeviceReading.isPresent()) {
-                        updateStatus(ThingStatus.ONLINE);
                         logger.debug("Successfully updated device {} readings! Now updating channels", getId());
                         DEVICE theDevice = newDeviceReading.get();
                         this.device = theDevice;
+                        updateStatus(isReachable() ? ThingStatus.ONLINE : ThingStatus.OFFLINE);
                         updateProperties(theDevice);
                         getDataTimestamp().ifPresent(dataTimeStamp -> {
                             strategy.setDataTimeStamp(dataTimeStamp, timeZoneProvider.getTimeZone());

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/NetatmoModuleHandler.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/NetatmoModuleHandler.java
@@ -99,7 +99,7 @@ public class NetatmoModuleHandler<MODULE> extends AbstractNetatmoThingHandler {
     protected void updateChannels(Object module) {
         MODULE theModule = (MODULE) module;
         setModule(theModule);
-        updateStatus(ThingStatus.ONLINE);
+        updateStatus(isReachable() ? ThingStatus.ONLINE : ThingStatus.OFFLINE);
         getRadioHelper().ifPresent(helper -> helper.setModule(module));
         getBatteryHelper().ifPresent(helper -> helper.setModule(module));
         updateProperties(theModule);

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/station/NAMainHandler.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/station/NAMainHandler.java
@@ -282,4 +282,15 @@ public class NAMainHandler extends NetatmoDeviceHandler<NAMain> {
     protected Optional<Integer> getDataTimestamp() {
         return getDevice().map(d -> d.getLastStatusStore());
     }
+
+    @Override
+    protected boolean isReachable() {
+        boolean result = false;
+        Optional<NAMain> device = getDevice();
+        if (device.isPresent()) {
+            Boolean reachable = device.get().getReachable();
+            result = reachable != null ? reachable.booleanValue() : false;
+        }
+        return result;
+    }
 }

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/station/NAModule1Handler.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/station/NAModule1Handler.java
@@ -18,6 +18,7 @@ import static org.openhab.binding.netatmo.internal.NetatmoBindingConstants.*;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.concurrent.ConcurrentHashMap;
 
 import org.eclipse.jdt.annotation.NonNullByDefault;
@@ -169,5 +170,16 @@ public class NAModule1Handler extends NetatmoModuleHandler<NAStationModule> {
         }
 
         return super.getNAThingProperty(channelId);
+    }
+
+    @Override
+    protected boolean isReachable() {
+        boolean result = false;
+        Optional<NAStationModule> module = getModule();
+        if (module.isPresent()) {
+            Boolean reachable = module.get().getReachable();
+            result = reachable != null ? reachable.booleanValue() : false;
+        }
+        return result;
     }
 }

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/station/NAModule2Handler.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/station/NAModule2Handler.java
@@ -15,6 +15,8 @@ package org.openhab.binding.netatmo.internal.station;
 import static org.openhab.binding.netatmo.internal.ChannelTypeUtils.*;
 import static org.openhab.binding.netatmo.internal.NetatmoBindingConstants.*;
 
+import java.util.Optional;
+
 import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.eclipse.smarthome.core.i18n.TimeZoneProvider;
 import org.eclipse.smarthome.core.thing.Thing;
@@ -64,5 +66,16 @@ public class NAModule2Handler extends NetatmoModuleHandler<NAStationModule> {
             }
         }
         return super.getNAThingProperty(channelId);
+    }
+
+    @Override
+    protected boolean isReachable() {
+        boolean result = false;
+        Optional<NAStationModule> module = getModule();
+        if (module.isPresent()) {
+            Boolean reachable = module.get().getReachable();
+            result = reachable != null ? reachable.booleanValue() : false;
+        }
+        return result;
     }
 }

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/station/NAModule3Handler.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/station/NAModule3Handler.java
@@ -18,6 +18,7 @@ import static org.openhab.binding.netatmo.internal.NetatmoBindingConstants.*;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.concurrent.ConcurrentHashMap;
 
 import org.eclipse.jdt.annotation.NonNullByDefault;
@@ -88,5 +89,16 @@ public class NAModule3Handler extends NetatmoModuleHandler<NAStationModule> {
         }
 
         return super.getNAThingProperty(channelId);
+    }
+
+    @Override
+    protected boolean isReachable() {
+        boolean result = false;
+        Optional<NAStationModule> module = getModule();
+        if (module.isPresent()) {
+            Boolean reachable = module.get().getReachable();
+            result = reachable != null ? reachable.booleanValue() : false;
+        }
+        return result;
     }
 }

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/station/NAModule4Handler.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/station/NAModule4Handler.java
@@ -18,6 +18,7 @@ import static org.openhab.binding.netatmo.internal.NetatmoBindingConstants.*;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.concurrent.ConcurrentHashMap;
 
 import org.eclipse.jdt.annotation.NonNullByDefault;
@@ -196,5 +197,16 @@ public class NAModule4Handler extends NetatmoModuleHandler<NAStationModule> {
         }
 
         return super.getNAThingProperty(channelId);
+    }
+
+    @Override
+    protected boolean isReachable() {
+        boolean result = false;
+        Optional<NAStationModule> module = getModule();
+        if (module.isPresent()) {
+            Boolean reachable = module.get().getReachable();
+            result = reachable != null ? reachable.booleanValue() : false;
+        }
+        return result;
     }
 }

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/welcome/NAWelcomeHomeHandler.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/welcome/NAWelcomeHomeHandler.java
@@ -30,7 +30,6 @@ import org.eclipse.smarthome.core.library.types.StringType;
 import org.eclipse.smarthome.core.thing.Thing;
 import org.eclipse.smarthome.core.types.State;
 import org.eclipse.smarthome.core.types.UnDefType;
-import org.eclipse.smarthome.io.net.http.HttpUtil;
 import org.openhab.binding.netatmo.internal.ChannelTypeUtils;
 import org.openhab.binding.netatmo.internal.camera.CameraHandler;
 import org.openhab.binding.netatmo.internal.handler.AbstractNetatmoThingHandler;
@@ -132,7 +131,7 @@ public class NAWelcomeHomeHandler extends NetatmoDeviceHandler<NAWelcomeHome> {
                     return UnDefType.UNDEF;
                 }
             case CHANNEL_WELCOME_EVENT_SNAPSHOT:
-                return findSnapshotURL().map(url -> (State) HttpUtil.downloadImage(url)).orElse(UnDefType.UNDEF);
+                return findSnapshotURL().map(url -> toRawType(url)).orElse(UnDefType.UNDEF);
             case CHANNEL_WELCOME_EVENT_SNAPSHOT_URL:
                 return findSnapshotURL().map(ChannelTypeUtils::toStringType).orElse(UnDefType.UNDEF);
             case CHANNEL_WELCOME_EVENT_VIDEO_URL:

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/welcome/NAWelcomePersonHandler.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/welcome/NAWelcomePersonHandler.java
@@ -27,7 +27,6 @@ import org.eclipse.smarthome.core.thing.Thing;
 import org.eclipse.smarthome.core.types.Command;
 import org.eclipse.smarthome.core.types.State;
 import org.eclipse.smarthome.core.types.UnDefType;
-import org.eclipse.smarthome.io.net.http.HttpUtil;
 import org.openhab.binding.netatmo.internal.handler.NetatmoBridgeHandler;
 import org.openhab.binding.netatmo.internal.handler.NetatmoModuleHandler;
 
@@ -76,6 +75,7 @@ public class NAWelcomePersonHandler extends NetatmoModuleHandler<NAWelcomePerson
     @Override
     protected State getNAThingProperty(String channelId) {
         Optional<NAWelcomeEvent> lastEvt = getLastEvent();
+        String url;
         switch (channelId) {
             case CHANNEL_WELCOME_PERSON_LASTSEEN:
                 return getModule().map(m -> toDateTimeType(m.getLastSeen(), timeZoneProvider.getTimeZone()))
@@ -87,7 +87,8 @@ public class NAWelcomePersonHandler extends NetatmoModuleHandler<NAWelcomePerson
             case CHANNEL_WELCOME_PERSON_AVATAR_URL:
                 return toStringType(getAvatarURL());
             case CHANNEL_WELCOME_PERSON_AVATAR:
-                return getAvatarURL() != null ? HttpUtil.downloadImage(getAvatarURL()) : UnDefType.UNDEF;
+                url = getAvatarURL();
+                return url != null ? toRawType(url) : UnDefType.UNDEF;
             case CHANNEL_WELCOME_PERSON_LASTMESSAGE:
                 return (lastEvt.isPresent() && lastEvt.get().getMessage() != null)
                         ? toStringType(lastEvt.get().getMessage().replace("<b>", "").replace("</b>", ""))
@@ -96,7 +97,8 @@ public class NAWelcomePersonHandler extends NetatmoModuleHandler<NAWelcomePerson
                 return lastEvt.isPresent() ? toDateTimeType(lastEvt.get().getTime(), timeZoneProvider.getTimeZone())
                         : UnDefType.UNDEF;
             case CHANNEL_WELCOME_PERSON_LASTEVENT:
-                return getLastEventURL() != null ? HttpUtil.downloadImage(getLastEventURL()) : UnDefType.UNDEF;
+                url = getLastEventURL();
+                return url != null ? toRawType(url) : UnDefType.UNDEF;
             case CHANNEL_WELCOME_PERSON_LASTEVENT_URL:
                 return getLastEventURL() != null ? toStringType(getLastEventURL()) : UnDefType.UNDEF;
         }

